Program security - Signer auth lesson updated (#432)
solana-foundation/developer-content: content/courses/program-security/signer-auth.md

---
title: Signer Authorization
objectives:
  - Explain the security risks of not performing appropriate signer checks.
  - Implement signer checks using native Rust
  - Implement signer checks using Anchor’s `Signer` type
  - Implement signer checks using Anchor’s `#[account(signer)]` constraint
description:
  "Ensure instructions are only executed by authorized accounts by implementing
  signer checks."
---

## Summary

- **Signer Checks** are essential to verify that specific accounts have signed a
  transaction. Without proper signer checks, unauthorized accounts may execute
  instructions they shouldn't be allowed to perform.

> - In Rust, implement a signer check by verifying that an account's `is_signer`

Reviewer: ```suggestion
- In native Rust, implement a signer check by verifying that an account's `is_signer`
```

Reviewer: We should probably start with Anchor, since that's the most common path.

Author: Updated content to start with Anchor.

  }
  ```

- In Anchor, you can use the `Signer` account type in your account validation
  struct to automatically perform a signer check on a given account.
- Anchor also provides the
  [`#[account(signer)]`](https://www.anchor-lang.com/docs/account-constraints)
  constraint, which automatically verifies that a specified account has signed
  the transaction.

## Lesson

**Signer checks** ensure that only authorized accounts can execute specific
instructions. Without these checks, any account might perform operations that
should be restricted, potentially leading to severe security vulnerabilities,
such as unauthorized access and control over program accounts.

### Missing Signer Check

Below is an oversimplified instruction handler that updates the `authority`
field on a program account. Notice that the `authority` field in the
`UpdateAuthority` account validation struct is of type `UncheckedAccount`. In
Anchor, the
[`UncheckedAccount`](https://docs.rs/anchor-lang/latest/anchor_lang/accounts/unchecked_account/struct.UncheckedAccount.html)
type indicates that no checks are performed on the account before executing the

> instruction handler.

Reviewer: I appreciate the attention to detail here.

}
```

### Adding Signer Authorization Checks

To validate that the `authority` account signed the transaction, add a signer
check within the instruction handler:

> ```rust

Reviewer: lol

      expect(error).to.be.an("error");
      console.log("Error message:", error.message);
    }
  });
})
```

Run `anchor test` to see that the transaction now returns a signature
verification error.

```bash
signer-authorization
Error message: Signature verification failed.
Missing signature for public key [`GprrWv9r8BMxQiWea9MrbCyK7ig7Mj8CcseEbJhDDZXM`].
    ✔ fails to perform secure withdraw with incorrect signer
```

This example shows how important it is to think through who should authorize
instructions and ensure that each is a signer on the transaction.

To review the final solution code, you can find it on the
[`solution` branch of the repository](https://github.com/solana-developers/solana-signer-auth/tree/solution).

## Challenge

Now that you've worked through the labs and challenges in this course, it's time
to apply your knowledge in a practical setting. For this challenge and those
that follow on security vulnerabilities, audit your own programs for the
specific vulnerability discussed in each lesson.

### Steps

1. **Audit Your Program or Find an Open Source Project**:

   - Begin by auditing your own code for missing signer checks, or find an open

>      source Solana program to audit. A great place to start is with the
>      [SPL programs](https://github.com/solana-labs/solana-program-library) if
>      you’re comfortable working with native Rust.

Reviewer: ```suggestion
     source Solana program to audit. A great place to start is with the
     [program examples](https://github.com/solana-developers/program-examples) repo.
```

Reviewer: This is a better codebase as it has anchor examples (and could use an audit!).

Author: Yes, I updated the link for the same.
